screwdriver/api/api.go

```diff
@@ -33,102 +36,136 @@ type ResponseError struct {
-	uri, err := url.Parse(c.baseURL)
+func (c client) GetCommand(smallSpec *util.CommandSpec) (*util.CommandSpec, error) {
+	uri, err := parseURL(c.baseURL)
 	if err != nil {
-		return nil, fmt.Errorf("The base Screwdriver API is invalid %q", c.baseURL)
+		return nil, fmt.Errorf("Failed to parse URL on GET: %v", err)
 	}
 	uri.Path = path.Join(uri.Path, "commands", smallSpec.Namespace, smallSpec.Name, smallSpec.Version)
-	req, err := http.NewRequest("GET", uri.String(), strings.NewReader(""))
-	if err != nil {
-		return nil, fmt.Errorf("Failed to create request about command to Screwdriver API: %v", err)
-	}
 
-	req.Header.Set("Content-type", "application/json")
-	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.jwt))
-	res, err := c.client.Do(req)
+	// Request to api
+	bodyBytes, statusCode, err := c.httpRequest("GET", uri.String(), c.jwt, "")
 	if err != nil {
-		return nil, fmt.Errorf("Failed to get command from Screwdriver API: %v", err)
+		return nil, fmt.Errorf("Failed to get http response: %v", err)
 	}
-	defer res.Body.Close()
-	body, err := handleResponse(res)
+
+	// Get response body
+	bodyBytes, err = handleResponse(bodyBytes, statusCode)
 
 	if err != nil {
 		return nil, err
 	}
 
-	err = json.Unmarshal(body, cmd)
+	cmd := new(util.CommandSpec)
+	err = json.Unmarshal(bodyBytes, cmd)
 	if err != nil {
 		return nil, fmt.Errorf("Failed to unmarshal command: %v", err)
 	}
 	return cmd, nil
 }
+
+func (c client) PostCommand(commandSpec *util.CommandSpec) error {
+	// Generate URL
+	uri, err := parseURL(c.baseURL)
+	if err != nil {
+		return fmt.Errorf("Failed to parse URL on POST: %v", err)
+	}
+	uri.Path = path.Join(uri.Path, "commands")
+
+	// Generate payload
+	commandSpecBytes, err := json.Marshal(&commandSpec)
+	var payload util.PostPayload
+	payload.Yaml = string(commandSpecBytes)
+	payloadBytes, err := json.Marshal(&payload)
+	if err != nil {
+		return fmt.Errorf("Failed to generate payload: %v", err)
+	}
+
+	bodyBytes, statusCode, err := c.httpRequest("POST", uri.String(), c.jwt, string(payloadBytes))
+	if err != nil {
+		return fmt.Errorf("Post request failed: %v", err)
+	}
+
+	bodyBytes, err = handleResponse(bodyBytes, statusCode)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func parseURL(urlstr string) (*url.URL, error) {
+	uri, err := url.Parse(urlstr)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to parse URL of Screwdriver API. URL is %v", urlstr)
+	}
+	return uri, nil
+}
+
+func (c client) httpRequest(method, url, token string, payload string) ([]byte, int, error) {
+	req, err := http.NewRequest(
+		method,
+		url,
+		bytes.NewBuffer([]byte(payload)),
+	)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, 0, fmt.Errorf("Faild to get http response: %v", err)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
```

Comment:
I think it is better to add error handling code for this method.

Reply:
You're right. I've added error handing.